data/transactions/logic/resources.go

```diff
@@ -0,0 +1,356 @@
+// apps, assets, and boxes that are available to a transaction, outside the
+// direct foreign array mechanism.
+type resources struct {
+	// These resources were created previously in the group, so they can be used
+	// by later transactions.
+	createdAsas []basics.AssetIndex
+	createdApps []basics.AppIndex
+
+	// These resources have been mentioned by some txn in the group, so they are
+	// available. But only their "main" data is available. For example, if an
+	// account is mentioned, its algo balance is available, but not necessarily
+	// its asset balance for any old ASA.
+	sharedAccounts map[basics.Address]struct{}
+	sharedAsas     map[basics.AssetIndex]struct{}
+	sharedApps     map[basics.AppIndex]struct{}
+	// We need to carefully track the "cross-product" availability, because if
+	// tx0 mentions an account A, and tx1 mentions an ASA X, that does _not_
+	// make the holding AX available
+	sharedHoldings map[ledgercore.AccountAsset]struct{}
+	sharedLocals   map[ledgercore.AccountApp]struct{}
+
+	// boxes are all of the top-level box refs from the txgroup. Most are added
+	// during NewEvalParams(). refs using 0 on an appl create are resolved and
+	// added when the appl executes. The boolean value indicates the "dirtiness"
+	// of the box - has it been modified in this txngroup? If yes, the size of
+	// the box counts against the group writeBudget. So delete is NOT a dirtying
+	// operation.
+	boxes map[boxRef]bool
+
+	// dirtyBytes maintains a running count of the number of dirty bytes in `boxes`
+	dirtyBytes uint64
+}
+
+func (r *resources) shareHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+}
+
+func (r *resources) shareAccountAndHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedAccounts[addr] = struct{}{}
+	if id != 0 {
+		r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+	}
+}
+
+func (r *resources) shareLocal(addr basics.Address, id basics.AppIndex) {
+	r.sharedLocals[ledgercore.AccountApp{Address: addr, App: id}] = struct{}{}
+}
+
+// In the fill* and allows* routines, we pass the header and the fields in
+// separately, even though they are pointers into the same structure. That
+// prevents dumb attempts to use other fields from the transaction.
+
+func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
+	switch tx.Type {
+	case protocol.PaymentTx:
+		r.fillPayment(&tx.Header, &tx.PaymentTxnFields)
+	case protocol.KeyRegistrationTx:
+		r.fillKeyRegistration(&tx.Header)
+	case protocol.AssetConfigTx:
+		r.fillAssetConfig(&tx.Header, &tx.AssetConfigTxnFields)
+	case protocol.AssetTransferTx:
+		r.fillAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		r.fillAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		r.fillApplicationCall(ep, &tx.Header, &tx.ApplicationCallTxnFields)
+	default:
```

Comment:
what about protocol.StateProofTx

Reply:
My first response: I am happy to let a panic kill such an attempt.  It would only happen if a user created a transaction group with an app call and a state proof txn, which they are clearly not allowed to do.  That's because we don't `fill` unless there is an app call in the group.  The standard state proof transaction is a singleton.

But I'm reconsidering. I suppose I'm _also_ ok with just doing nothing if this code encounters a state proof txn.  It's not really this func's job to panic in that situation.

That was my original thinking when I simply did nothing in the default case. Then, any new transaction type simply wouldn't add to the availability set.  I think @jasonpaulos felt we should complain in the case of a new txn type, and I felt state proof txn was essentially in the same category.

With all that in mind, what do the two of you think?

Reply:
Hmm, I think I'm team no-op, because it is a defined transaction type, it's the only one you don't have, and the type has no affect on the results.  otherwise it looks like something you forgot.  Or if you want to panic, call it out explicitly as a case.

Reply:
Explicitly adding a case for state proof txns which is a no-op seems fine to me. My main reason for suggesting panicking was to catch any new txn types in the future -- with a panic, we will be forced to remember this decision point and do something intentional here.

Reply:
ok, I'll go with an explicit no-op case for state proof, panic for a new type.